From 347058b5d173023aadfde0ef0cbac82b2e13c4a5 Mon Sep 17 00:00:00 2001
From: Jakub Hrozek <jhrozek@redhat.com>
Date: Thu, 24 Jul 2014 11:46:40 +0200
Subject: [PATCH] host_callback: Fall back to AF_INET on searching with
 AF_UNSPEC

Previously, when an ares_gethostbyname() searched with AF_UNSPEC and the
first AF_INET6 call only returned CNAMEs, the host_callback never
retried AF_INET.

This patch makes sure than on ARES_SUCCESS, the result of AF_INET6 is
taken as authoritative only if the result contains some addresses.
---
 ares_gethostbyname.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/ares_gethostbyname.c b/ares_gethostbyname.c
index 2b27b2e1048fccc0d4b0b9b8863bae64f86a1b6e..ba6b0f0f0c56615d8ec2016e42adccad79f65a2d 100644
--- a/ares_gethostbyname.c
+++ b/ares_gethostbyname.c
@@ -188,8 +188,9 @@ static void host_callback(void *arg, int status, int timeouts,
       else if (hquery->sent_family == AF_INET6)
         {
           status = ares_parse_aaaa_reply(abuf, alen, &host, NULL, NULL);
-          if ((status == ARES_ENODATA || status == ARES_EBADRESP) &&
-               hquery->want_family == AF_UNSPEC) {
+          if ((status == ARES_ENODATA || status == ARES_EBADRESP ||
+               (status == ARES_SUCCESS && host && host->h_addr_list[0] == NULL)) &&
+                hquery->want_family == AF_UNSPEC) {
             /* The query returned something but either there were no AAAA
                records (e.g. just CNAME) or the response was malformed.  Try
                looking up A instead. */
-- 
1.9.3

